File: datalog/skill/tools/ts-facts/test/fixtures/refs/src/use.ts
import { Base, Circle, Plain, type Shape } from "./shapes.js";

function logged(value: Function, context: ClassMethodDecoratorContext): void {
  void value;
  void context;
}

export class Registry {
  items: Shape[] = [];
  handler = (s: Shape): number => s.area();

  @logged
  add(s: Shape): void {
    this.items.push(s);
  }
}

export function total(shapes: Shape[], each: (s: Shape) => number): number {
  let sum = 0;
  for (const s of shapes) sum += each(s);
  return sum;
}

export function main(): string {
  const c = new Circle(2);
  c.grow(1);
  const r = new Registry();
  r.add(c);
  r.handler(c);
  new Plain();
  const b: Base = Base.create();
  total([b], (s) => s.area());
  return b.describe();
}

File: code-analysis/tools/code-facts/test/fixtures/java-refs/app/Use.java
package app;

import java.io.IOException;
import java.util.List;
import java.util.function.Supplier;
import shapes.Base;
import shapes.Circle;
import shapes.Plain;
import shapes.Shape;

public class Use {
  List<Shape> shapes;

  double total(List<Shape> items) {
    double sum = 0;
    for (Shape s : items) sum += s.area();
    items.forEach(s -> s.area());
    return sum;
  }

  Base make() throws IOException, IllegalStateException {
    Circle c = new Circle(2);
    c.radius = 3;
    new Plain();
    Supplier<Circle> fresh = Circle::new;
    Shape anon = new Shape() {
      @Override
      public double area() {
        return 0;
      }
    };
    if (anon instanceof Circle) return (Circle) anon;
    return Base.unit();
  }
}
